File: gateway/src/main/java/com/me/webdesigner/util/ViewValidator.java
package com.me.webdesigner.util;

import com.inductiveautomation.ignition.common.gson.JsonArray;
import com.inductiveautomation.ignition.common.gson.JsonElement;
import com.inductiveautomation.ignition.common.gson.JsonObject;
import com.inductiveautomation.ignition.common.gson.JsonPrimitive;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * View and Component Validation Utility
 *
 * Validates view.json structure and component hierarchy to ensure
 * data integrity before saving.
 *
 * Features:
 * - View structure validation
 * - Component hierarchy validation
 * - Required field validation
 * - Type checking
 * - Circular reference detection
 * - Duplicate name detection
 * - Size limits
 *
 * Usage:
 * ```java
 * ValidationResult result = ViewValidator.validateView(viewJson);
 * if (!result.isValid()) {
 *     // Handle errors
 *     for (String error : result.getErrors()) {
 *         logger.error(error);
 *     }
 * }
 * ```
 */
public final class ViewValidator {

    // Configuration
    private static final int MAX_NESTING_DEPTH = 20;
    private static final int MAX_COMPONENT_COUNT = 500;
    private static final int MAX_NAME_LENGTH = 100;

    // Required view fields
    private static final String[] REQUIRED_VIEW_FIELDS = {"root"};

    // Required component fields
    private static final String[] REQUIRED_COMPONENT_FIELDS = {"type"};

    // Private constructor - utility class
    private ViewValidator() {}

    /**
     * Validation result container
     */
    public static class ValidationResult {
        private final boolean valid;
        private final List<String> errors;
        private final List<String> warnings;

        public ValidationResult() {
            this.valid = true;
            this.errors = new ArrayList<>();
            this.warnings = new ArrayList<>();
        }

        public ValidationResult(List<String> errors, List<String> warnings) {
            this.valid = errors.isEmpty();
            this.errors = errors;
            this.warnings = warnings;
        }

        public boolean isValid() {
            return valid;
        }

        public List<String> getErrors() {
            return errors;
        }

        public List<String> getWarnings() {
            return warnings;
        }

        public boolean hasWarnings() {
            return !warnings.isEmpty();
        }
    }

    /**
     * Validate a complete view.json structure
     */
    public static ValidationResult validateView(JsonObject viewJson) {
        List<String> errors = new ArrayList<>();
        List<String> warnings = new ArrayList<>();

        // Check null
        if (viewJson == null) {
            errors.add("View JSON cannot be null");
            return new ValidationResult(errors, warnings);
        }

        // Validate required fields
        for (String field : REQUIRED_VIEW_FIELDS) {
            if (!viewJson.has(field)) {
                errors.add("Missing required field: " + field);
            }
        }

        if (!errors.isEmpty()) {
            return new ValidationResult(errors, warnings);
        }

        // Validate root component
        JsonElement rootElement = viewJson.get("root");
        if (!rootElement.isJsonObject()) {
            errors.add("Root must be a JSON object");
            return new ValidationResult(errors, warnings);
        }

        JsonObject root = rootElement.getAsJsonObject();

        // Validate component hierarchy
        Set<String> componentNames = new HashSet<>();
        int[] componentCount = {0};

        validateComponent(root, 0, componentNames, componentCount, errors, warnings);

        // Check component count limit
        if (componentCount[0] > MAX_COMPONENT_COUNT) {
            errors.add(String.format("Too many components (%d). Maximum allowed: %d",
                componentCount[0], MAX_COMPONENT_COUNT));
        }

        return new ValidationResult(errors, warnings);
    }

    /**
     * Validate a single component and its children recursively
     */
    private static void validateComponent(JsonObject component, int depth,
                                         Set<String> componentNames,
                                         int[] componentCount,
                                         List<String> errors,
                                         List<String> warnings) {
        // Increment counter
        componentCount[0]++;

        // Check nesting depth
        if (depth > MAX_NESTING_DEPTH) {
            errors.add(String.format("Component nesting too deep (depth: %d). Maximum allowed: %d",
                depth, MAX_NESTING_DEPTH));
            return;
        }

        // Validate required fields
        for (String field : REQUIRED_COMPONENT_FIELDS) {
            if (!component.has(field)) {
                errors.add(String.format("Component at depth %d missing required field: %s", depth, field));
                return;
            }
        }

        // Validate type field
        JsonElement typeElement = component.get("type");
        if (!typeElement.isJsonPrimitive() || !typeElement.getAsJsonPrimitive().isString()) {
            errors.add(String.format("Component at depth %d has invalid type (must be string)", depth));
            return;
        }

        String type = typeElement.getAsString();
        if (type == null || type.trim().isEmpty()) {
            errors.add(String.format("Component at depth %d has empty type", depth));
            return;
        }

        // Validate meta.name if present
        if (component.has("meta")) {
            JsonElement metaElement = component.get("meta");
            if (metaElement.isJsonObject()) {
                JsonObject meta = metaElement.getAsJsonObject();
                if (meta.has("name")) {
                    JsonElement nameElement = meta.get("name");
                    if (nameElement.isJsonPrimitive() && nameElement.getAsJsonPrimitive().isString()) {
                        String name = nameElement.getAsString();

                        // Check name length
                        if (name.length() > MAX_NAME_LENGTH) {
                            errors.add(String.format("Component name '%s' exceeds maximum length (%d)",
                                name, MAX_NAME_LENGTH));
                        }

                        // Check for duplicate names
                        if (componentNames.contains(name)) {
                            warnings.add(String.format("Duplicate component name: %s", name));
                        } else {
                            componentNames.add(name);
                        }
                    }
                }
            }
        }

        // Validate children if present
        if (component.has("children")) {
            JsonElement childrenElement = component.get("children");
            if (!childrenElement.isJsonArray()) {
                errors.add(String.format("Component '%s' has invalid children (must be array)", type));
                return;
            }

            JsonArray children = childrenElement.getAsJsonArray();
            for (int i = 0; i < children.size(); i++) {
                JsonElement childElement = children.get(i);
                if (!childElement.isJsonObject()) {
                    errors.add(String.format("Child %d of component '%s' is not an object", i, type));
                    continue;
                }

                validateComponent(childElement.getAsJsonObject(), depth + 1,
                    componentNames, componentCount, errors, warnings);
            }
        }

        // Validate layout if present
        if (component.has("layout")) {
            validateLayout(component.get("layout"), type, errors, warnings);
        }

        // Validate props if present
        if (component.has("props")) {
            JsonElement propsElement = component.get("props");
            if (!propsElement.isJsonObject()) {
                warnings.add(String.format("Component '%s' has invalid props (expected object)", type));
            }
        }
    }

    /**
     * Validate layout object
     */
    private static void validateLayout(JsonElement layoutElement, String componentType,
                                      List<String> errors, List<String> warnings) {
        if (!layoutElement.isJsonObject()) {
            warnings.add(String.format("Component '%s' has invalid layout (must be object)", componentType));
            return;
        }

        JsonObject layout = layoutElement.getAsJsonObject();

        // Validate numeric fields
        String[] numericFields = {"x", "y", "width", "height"};
        for (String field : numericFields) {
            if (layout.has(field)) {
                JsonElement fieldElement = layout.get(field);
                if (!fieldElement.isJsonPrimitive()) {
                    warnings.add(String.format("Component '%s' layout.%s must be a number", componentType, field));
                    continue;
                }

                JsonPrimitive primitive = fieldElement.getAsJsonPrimitive();
                if (!primitive.isNumber()) {
                    warnings.add(String.format("Component '%s' layout.%s must be a number", componentType, field));
                }
            }
        }

        // Validate width and height are positive
        if (layout.has("width")) {
            try {
                int width = layout.get("width").getAsInt();
                if (width < 0) {
                    warnings.add(String.format("Component '%s' has negative width", componentType));
                }
            } catch (Exception e) {
                // Already caught above
            }
        }

        if (layout.has("height")) {
            try {
                int height = layout.get("height").getAsInt();
                if (height < 0) {
                    warnings.add(String.format("Component '%s' has negative height", componentType));
                }
            } catch (Exception e) {
                // Already caught above
            }
        }
    }

    /**
     * Quick validation check - just checks critical errors
     */
    public static boolean isValidView(JsonObject viewJson) {
        ValidationResult result = validateView(viewJson);
        return result.isValid();
    }

    /**
     * Validate component type format
     */
    public static boolean isValidComponentType(String type) {
        if (type == null || type.trim().isEmpty()) {
            return false;
        }

        // Component types should follow format: namespace.category.name
        // Examples: ia.display.label, ia.input.button
        String[] parts = type.split("\\.");
        return parts.length >= 2;
    }

    /**
     * Sanitize component name
     */
    public static String sanitizeComponentName(String name) {
        if (name == null) {
            return "";
        }

        // Remove invalid characters
        String sanitized = name.replaceAll("[^a-zA-Z0-9_-]", "_");

        // Trim to max length
        if (sanitized.length() > MAX_NAME_LENGTH) {
            sanitized = sanitized.substring(0, MAX_NAME_LENGTH);
        }

        return sanitized;
    }
}

File: frontend/src/utils/viewValidator.ts
/**
 * View and Component Validation Utility (Frontend)
 *
 * Validates view structures and component hierarchy on the client side
 * before sending to the backend.
 *
 * Features:
 * - View structure validation
 * - Component validation
 * - Type checking
 * - Required field validation
 * - Size limits
 * - Circular reference detection
 * - Duplicate name detection
 */

export interface ValidationResult {
  valid: boolean
  errors: string[]
  warnings: string[]
}

export interface Component {
  type: string
  meta?: {
    name?: string
  }
  props?: Record<string, any>
  layout?: {
    x?: number
    y?: number
    width?: number
    height?: number
  }
  children?: Component[]
  [key: string]: any
}

export interface View {
  root: Component
  params?: Record<string, any>
  custom?: Record<string, any>
}

// Configuration
const MAX_NESTING_DEPTH = 20
const MAX_COMPONENT_COUNT = 500
const MAX_NAME_LENGTH = 100
const MIN_SIZE = 1
const MAX_SIZE = 10000

/**
 * Validate a complete view structure
 */
export function validateView(view: any): ValidationResult {
  const errors: string[] = []
  const warnings: string[] = []

  // Check null/undefined
  if (!view) {
    errors.push('View cannot be null or undefined')
    return { valid: false, errors, warnings }
  }

  // Check required fields
  if (!view.root) {
    errors.push('View must have a root component')
    return { valid: false, errors, warnings }
  }

  // Validate root component
  const componentNames = new Set<string>()
  let componentCount = 0

  validateComponent(view.root, 0, componentNames, componentCount, errors, warnings, (count) => {
    componentCount = count
  })

  // Check component count limit
  if (componentCount > MAX_COMPONENT_COUNT) {
    errors.push(`Too many components (${componentCount}). Maximum allowed: ${MAX_COMPONENT_COUNT}`)
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  }
}

/**
 * Validate a single component and its children
 */
function validateComponent(
  component: any,
  depth: number,
  componentNames: Set<string>,
  componentCount: number,
  errors: string[],
  warnings: string[],
  updateCount: (count: number) => void
): void {
  // Increment counter
  componentCount++
  updateCount(componentCount)

  // Check nesting depth
  if (depth > MAX_NESTING_DEPTH) {
    errors.push(`Component nesting too deep (depth: ${depth}). Maximum allowed: ${MAX_NESTING_DEPTH}`)
    return
  }

  // Check if component is an object
  if (!component || typeof component !== 'object') {
    errors.push(`Component at depth ${depth} is not an object`)
    return
  }

  // Validate required fields
  if (!component.type) {
    errors.push(`Component at depth ${depth} missing required field: type`)
    return
  }

  // Validate type field
  if (typeof component.type !== 'string' || component.type.trim() === '') {
    errors.push(`Component at depth ${depth} has invalid type`)
    return
  }

  const type = component.type

  // Validate type format (should be namespace.category.name)
  if (!isValidComponentType(type)) {
    warnings.push(`Component type '${type}' does not follow standard format (namespace.category.name)`)
  }

  // Validate meta.name if present
  if (component.meta && component.meta.name) {
    const name = component.meta.name

    if (typeof name !== 'string') {
      warnings.push(`Component name must be a string, got ${typeof name}`)
    } else {
      // Check name length
      if (name.length > MAX_NAME_LENGTH) {
        errors.push(`Component name '${name}' exceeds maximum length (${MAX_NAME_LENGTH})`)
      }

      // Check for duplicate names
      if (componentNames.has(name)) {
        warnings.push(`Duplicate component name: ${name}`)
      } else {
        componentNames.add(name)
      }

      // Check name format
      if (!/^[a-zA-Z0-9_-]+$/.test(name)) {
        warnings.push(`Component name '${name}' contains invalid characters. Use only letters, numbers, underscores, and hyphens.`)
      }
    }
  }

  // Validate children if present
  if (component.children !== undefined) {
    if (!Array.isArray(component.children)) {
      errors.push(`Component '${type}' has invalid children (must be array)`)
    } else {
      component.children.forEach((child: any, index: number) => {
        if (!child || typeof child !== 'object') {
          errors.push(`Child ${index} of component '${type}' is not an object`)
        } else {
          validateComponent(child, depth + 1, componentNames, componentCount, errors, warnings, updateCount)
        }
      })
    }
  }

  // Validate layout if present
  if (component.layout) {
    validateLayout(component.layout, type, errors, warnings)
  }

  // Validate props if present
  if (component.props !== undefined) {
    if (typeof component.props !== 'object' || Array.isArray(component.props)) {
      warnings.push(`Component '${type}' has invalid props (expected object)`)
    }
  }
}

/**
 * Validate layout object
 */
function validateLayout(
  layout: any,
  componentType: string,
  errors: string[],
  warnings: string[]
): void {
  if (!layout || typeof layout !== 'object') {
    warnings.push(`Component '${componentType}' has invalid layout (must be object)`)
    return
  }

  // Validate numeric fields
  const numericFields = ['x', 'y', 'width', 'height']

  numericFields.forEach((field) => {
    if (field in layout) {
      const value = layout[field]

      if (typeof value !== 'number' || isNaN(value)) {
        warnings.push(`Component '${componentType}' layout.${field} must be a number`)
      } else {
        // Check width and height are positive
        if ((field === 'width' || field === 'height') && value < MIN_SIZE) {
          warnings.push(`Component '${componentType}' has invalid ${field}: ${value} (minimum: ${MIN_SIZE})`)
        }

        // Check maximum size
        if (value > MAX_SIZE) {
          warnings.push(`Component '${componentType}' layout.${field} exceeds maximum (${MAX_SIZE})`)
        }
      }
    }
  })
}

/**
 * Quick validation check - just checks critical errors
 */
export function isValidView(view: any): boolean {
  const result = validateView(view)
  return result.valid
}

/**
 * Validate component type format
 */
export function isValidComponentType(type: string): boolean {
  if (!type || typeof type !== 'string' || type.trim() === '') {
    return false
  }

  // Component types should follow format: namespace.category.name
  // Examples: ia.display.label, ia.input.button
  const parts = type.split('.')
  return parts.length >= 2
}

/**
 * Sanitize component name
 */
export function sanitizeComponentName(name: string): string {
  if (!name) {
    return ''
  }

  // Remove invalid characters
  let sanitized = name.replace(/[^a-zA-Z0-9_-]/g, '_')

  // Trim to max length
  if (sanitized.length > MAX_NAME_LENGTH) {
    sanitized = sanitized.substring(0, MAX_NAME_LENGTH)
  }

  return sanitized
}

/**
 * Validate property value by type
 */
export function validatePropertyValue(value: any, expectedType?: string): { valid: boolean; error?: string } {
  // Allow null/undefined
  if (value === null || value === undefined) {
    return { valid: true }
  }

  // If no expected type, allow any value
  if (!expectedType) {
    return { valid: true }
  }

  // Type-specific validation
  switch (expectedType.toLowerCase()) {
    case 'string':
      if (typeof value !== 'string') {
        return { valid: false, error: `Expected string, got ${typeof value}` }
      }
      break

    case 'number':
    case 'int':
    case 'integer':
    case 'float':
    case 'double':
      if (typeof value !== 'number' || isNaN(value)) {
        return { valid: false, error: `Expected number, got ${typeof value}` }
      }
      break

    case 'boolean':
    case 'bool':
      if (typeof value !== 'boolean') {
        return { valid: false, error: `Expected boolean, got ${typeof value}` }
      }
      break

    case 'object':
      if (typeof value !== 'object' || Array.isArray(value)) {
        return { valid: false, error: `Expected object, got ${typeof value}` }
      }
      break

    case 'array':
      if (!Array.isArray(value)) {
        return { valid: false, error: `Expected array, got ${typeof value}` }
      }
      break

    default:
      // Unknown type, allow it
      return { valid: true }
  }

  return { valid: true }
}

/**
 * Check if view has unsaved changes by comparing with original
 */
export function hasUnsavedChanges(current: any, original: any): boolean {
  return JSON.stringify(current) !== JSON.stringify(original)
}

/**
 * Deep clone a view or component
 */
export function cloneView(view: any): any {
  return JSON.parse(JSON.stringify(view))
}

/**
 * Check for circular references in view structure
 */
export function hasCircularReferences(obj: any, seen = new WeakSet()): boolean {
  if (obj === null || typeof obj !== 'object') {
    return false
  }

  if (seen.has(obj)) {
    return true
  }

  seen.add(obj)

  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      if (hasCircularReferences(obj[key], seen)) {
        return true
      }
    }
  }

  return false
}
